Support large metrics numbers

williamlardier opened the pull request:
We use strings because the usual number type is too limited (up to ~9PB). using string is an efficient way to transmit the information, and later, in nodejs, to parse it as BigInt.

Seragonia committed:
Return string metrics

Returning actual number won't work if they exceed the Max Safe
Integer, which is aroung 9PB of "totalBytes", for example.

Issue: SCUBA-217

Seragonia committed:
Update tests

- Test with big numbers.
- Make the tests independent from order of execution.

Issue: SCUBA-217

Seragonia committed:
Bump scubaclient version

Issue: SCUBA-217

francoisferrand commented on src/client.ts:
```diff
@@ -36,8 +36,8 @@ export type ScubaClientParameters = Omit<
-    objectsTotal: number;
-    bytesTotal: number;
+    objectsTotal: string;
+    bytesTotal: string;
```
should we not use BigInt in the (typescript) api, so that it is easy to consume?
and the client can then handle the (de)serialization to/from string when needed (and support numbers as well, for backwards compatibility in case it is used with an older scuba)

Seragonia committed:
Use ecma version 2020

Needed to know how to use BigInt.

Issue: SCUBA-217

Seragonia committed:
Make scubaclient return bigints

- It won't return string, this will be handled by the program
  using the values.

Issue: SCUBA-217

francoisferrand commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "scubaclient",
-  "version": "1.0.0",
+  "version": "1.0.1",
```
`1.1.0` ?

williamlardier replied:
yes!

williamlardier replied:
Done

francoisferrand commented on src/client.ts:
```diff
@@ -216,7 +220,21 @@ export default class ScubaClient {
+                    return;
+                }
+                // eslint-disable-next-line no-param-reassign
+                metric.bytesTotal = BigInt(metric.bytesTotal || 0);
```
if it were (untyped) javascript, this would be consistent; however, here we kind of break/abuse the typing system : at line 223, we say the response contains a metrics with `bigint` parameters.... so this line would be (according to the type-system) useless

* Is typescript allowed to optimize away such seamlingly useless conversions? If it is, we have an issue
* The cleaner way may be to have 2 types (one "internal" with strings, and another for the API) ; or to tweak the JSON deserialization, maybe with a "[reviver](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#using_the_reviver_parameter)" (c.f. https://stackoverflow.com/a/69644630)

williamlardier replied:
To me, it's more that we use the same `GetMetricsBatchResponse` type for both the received response and the function return type, while it should be two different types: we receive strings, and we return bigints.

I am against revivers, it slows down serialization and may lead to problem if we were to rely on it and eventually get BigInt in places we are not expecting them (because if we use revivers here, we should to it everywhere).

williamlardier replied:
As long as we control the objects we get, there is no point using reviver, better convert it explicitly when needed. I will update the types to differentiate between the received string and the returned bigint.

williamlardier replied:
Done.

Seragonia committed:
Use distinct types when handling scuba responses and return values

Issue: SCUBA-217

francoisferrand commented on src/client.ts:
```diff
@@ -35,9 +35,20 @@ export type ScubaClientParameters = Omit<
 };
 
+// API response types (with strings)
+export type ScubaReturnedMetrics = {
```
nit: better convey this is the "wire" format : `ScubaMetricsMessage` or `ScubaMetricsResponse` ? or something like `ScubaMetricsApi` vs `ScubeMetricsJson`...

(ideally make it consistency between all types)

williamlardier replied:
I will use `ScubaMetricsResponse`, which is consistent with all other types: sometimes the client would also return the "Response" type, which is fine as long as we don't modify it, we can create other types later if we were to add more logic.

francoisferrand commented on src/client.ts:
```diff
@@ -216,8 +251,34 @@ export default class ScubaClient {
+                if ('error' in resource) {
+                    return resource as GetMetricsBatchResponseResourceError;
+                }
+                return {
```
nit: to make the code more readable, this "conversion" could be a method of `GetMetricsBatchScubaResponseResourceMetrics` which does the conversion

williamlardier replied:
I am removing the "as": it is not needed, TS is able to find the right type thanks to the if check already.

Seragonia committed:
Do not rely on toJSON

- We were passing a test on serialize() thank to the global
  toJSON defined in the test, yet during execution time, we
  would not be able to serialize. We cannot use toJSON globally
  in this repo, because it would affect all projects importing
  it, and eventually lead to unexpected side effects.

Issue: SCUBA-217

Seragonia committed:
Improve types

Issue: SCUBA-217

ghost commented:
This PR was updated not to rely on toJSON: details in the commits. No logic was changed, but we explicitly convert to string when appropriate now. (in tests).
